[BE] Upload user avatar

MihaiDruta opened the pull request:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

1.Created Image entity.
2.One-to-one relation of user to image (avatar).
3.Added possibility to upload user avatar.
4.User avatar is center-cropped to a square.
5.Avatar resized to 256x256.
6.Added possibility to remove user avatar.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

1.Created Image entity.
2.One-to-one relation of user to image (avatar).
3.Added possibility to upload user avatar.
4.User avatar is center-cropped to a square.
5.Avatar resized to 256x256.
6.Added possibility to remove user avatar.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Original and resized avatars (256x256, 40x40) are now stored all in different directories.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Added php-imagick dependency to dockerfile.

SpartakusMd commented on config/services.yaml:
```diff
@@ -5,6 +5,7 @@
 # https://symfony.com/doc/current/best_practices/configuration.html#application-related-configuration
 parameters:
   locale: 'en'
+  user_image_directory: '%kernel.project_dir%/src/Images/User/'
```
This should be an in the `public` folder, not `src`. `src` is not available publicly.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -196,6 +199,17 @@ public function editUser(
 
         $data = $request->getContent();
 
+        $userAvatar = new UploadedBase64EncodedFile(new Base64EncodedFile($request->get('avatar')));
```
What if the user didn't change `avatar`?

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -513,4 +527,87 @@ public function getUserList(
+            ], Response::HTTP_FORBIDDEN);
+        }
+
+        if (!$user->getAvatar()) {
```
No error should be returned if the avatar is not present. Just a successful message.

SpartakusMd commented on src/Repository/ImageRepository.php:
```diff
@@ -0,0 +1,104 @@
+        $filename = $user->getId() . '.' . $uploadedFile->guessExtension();
+
+        $resizedImage256 = $this->cropperResizer->avatarCropResize($uploadedFile, 256);
+        $resizedImage256->move($this->userTargetDirectory . '256x256/', $filename);
```
The repository should be used only for database queries. Move this logic to a separate service.

SpartakusMd commented on src/Repository/ImageRepository.php:
```diff
@@ -0,0 +1,104 @@
+        $resizedImage256->move($this->userTargetDirectory . '256x256/', $filename);
+
+        $resizedImage40 = $this->cropperResizer->avatarCropResize($uploadedFile, 40);
+        $resizedImage40->move($this->userTargetDirectory . '40x40/', $filename);
```
Sizes should be configurable in a yaml file from `config/` folder.

SpartakusMd commented on src/Repository/ImageRepository.php:
```diff
@@ -0,0 +1,104 @@
+    private $cropperResizer;
+
+    public function __construct(
+        $userTargetDirectory,
```
Type string. This is not mandatory linked to user so you have to rename the variable.

SpartakusMd commented on src/Service/ImageCropperResizer.php:
```diff
@@ -0,0 +1,33 @@
+
+class ImageCropperResizer
+{
+    public function avatarCropResize(UploadedFile $userAvatar, int $size): UploadedBase64EncodedFile
```
If the class name is general, use general variables. Not `avatarCropResize()` but `cropAndResize()`. Not `$userAvatar` but `$uploadedFile`.

SpartakusMd replied:
Also maybe I need a diferent width than height?

SpartakusMd commented on src/Transformer/UserTransformer.php:
```diff
@@ -106,6 +120,15 @@ public function editTransform(
             }
         }
+
+        $userAvatar = new UploadedBase64EncodedFile(new Base64EncodedFile($dto->avatar));
```
What if no new avatar was sent?

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

1.Added avatar size and thumbnail size in services.yaml (sent in function as constructor parameters)
2.Verified if the avatar is uploaded or not and handling this situation properly.
3.Images are now stored in public directory
4.Created UserAvatarManager class who takes care of writing and removing images from directories
5.Generalized crop and resize functions in order to be used in another further situations
6.Updated .gitignore file

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Changed some variable names.

SpartakusMd commented on config/services.yaml:
```diff
@@ -5,6 +5,9 @@
 parameters:
   locale: 'en'
+  user_image_directory: '%kernel.project_dir%/public/Images/User/'
+  user_avatar_size: 200
```
Please use YAML arrays to specify the resolutions.

SpartakusMd commented on src/Repository/ImageRepository.php:
```diff
@@ -0,0 +1,80 @@
+
+        $this->directoryManager->saveAvatarInDirectory($uploadedFile, $user);
+        $image->setFile($filename);
+        $image->setAlt('User avatar');
```
Use the name of the user.

SpartakusMd commented on src/Repository/ImageRepository.php:
```diff
@@ -0,0 +1,80 @@
+        $image = new Image();
+        $filename = $user->getId() . '.' . $uploadedFile->guessExtension();
+
+        $this->directoryManager->saveAvatarInDirectory($uploadedFile, $user);
```
The repository is responsible only for interacting with the database. Building the `$image` should happen in other place.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -196,6 +199,19 @@ public function editUser(
 
         $data = $request->getContent();
 
+        if (!empty($request->get('avatar'))) {
```
Why is this in Controller? If it's only for validation move it to `UserTransformer`.

SpartakusMd replied:
THis was not tackled. The whole block should go into `UserTransformer`.

SpartakusMd commented on src/Repository/ImageRepository.php:
```diff
@@ -0,0 +1,80 @@
+    public function __construct(
+        RegistryInterface $registry,
+        UserRepository $userRepository,
+        UserAvatarManager $directoryManager
```
Why `UserAvatarManager` is named `directoryManager`?

SpartakusMd commented on src/Repository/UserRepository.php:
```diff
@@ -115,4 +116,18 @@ public function getPaginatedUserList(
+     * @throws ORMException
+     * @throws OptimisticLockException
+     */
+    public function saveAvatar(User $user, Image $avatar): void
```
Why another method? Instead of `$this->userRepository->saveAvatar($user, $image);` you can use:
```
$user->setAvatar($image);
$this->userRepository->save($user);
```

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

1.Used YAML array for specifying the resolutions.
2.Image alt is now the username.
3.Moved creating image function from repository.
4.Moved filetype validation to DTO as a function.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Fixed the situation when an user could send a blank field when changing his password.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Fixed duplicate operation in last migration. This change (CHANGE password_changed_at password_changed_at DATETIME NOT NULL) has been done in the previous migration.

SpartakusMd commented on src/Migrations/Version20190722112440.php:
```diff
@@ -24,7 +24,7 @@ public function up(Schema $schema): void
         $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql',
             'Migration can only be executed safely on \'mysql\'.');
 
-        $this->addSql('ALTER TABLE user ADD biography LONGTEXT DEFAULT NULL, CHANGE password_changed_at password_changed_at DATETIME NOT NULL');
```
DO NOT MODIFY existing migration !!! Create a new one.

SpartakusMd commented on src/Repository/ImageRepository.php:
```diff
@@ -0,0 +1,78 @@
+        User $user
+    ): void {
+        $image = $this->userAvatarManager->createImage($uploadedFile, $user);
+        $this->userAvatarManager->saveAvatarInDirectory($uploadedFile, $user);
```
ImageRepository is still doing extra work. Only read/write/save to database should be done here !!!

SpartakusMd commented on src/Service/UserAvatarManager.php:
```diff
@@ -0,0 +1,122 @@
+    ) {
+        $this->cropperResizer = $cropperResizer;
+        $this->targetDirectory = $targetDirectory;
+        $this->profilePicWidth = $avatarSize['profile']['width'];
```
You should loop over the `$avatarSize` so that it's not necessary to modify the code if you need a new size.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Fixed migrations

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Removed redundant logic from ImageRepository.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Parsing array from services.yaml using foreach for retrieving image dimensions.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

1.Created general TransformerException and NotValidFileType which extends TransformerException
2.Using sprintf() function to make the path of file more readable(for moving files to directory/removing them)
3.Removed redundant logic from repository(only save() and delete() functions remained)

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -234,20 +224,29 @@ public function editUser(
-                [
-                    'code' => Response::HTTP_NOT_FOUND,
+        } catch (TransformerException $exception) {
+            if ($exception instanceof EntityNotFound) {
```
If you handle the exceptions differently, use multiple `catch{}` blocks and each can have different exceptions to catch.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

Removed unused function from UserDTO.

SpartakusMd commented on src/Service/UserAvatarManager.php:
```diff
@@ -51,9 +51,10 @@ public function saveAvatarInDirectory(UploadedFile $uploadedFile, User $user): v
         foreach ($this->avatarSize as $size) {
             $image = $this->cropperResizer->resize($croppedImage, $size['width'], $size['height']);
-            $image->move($this->targetDirectory . $size['width'] . 'x' . $size['height'] . '/', $filename);
+            $image->move(sprintf($path, $this->targetDirectory, $size['width'], $size['height']), $filename);
```
Extract whole `sprintf()` to a variable.

SpartakusMd commented on src/Service/UserAvatarManager.php:
```diff
@@ -65,13 +66,9 @@ public function removeAvatarFromDirectory(User $user): void
-                . $size['height'] . '/'
-                . $filename
-            );
+            $filesystem->remove(sprintf($path, $this->targetDirectory, $size['width'], $size['height'], $filename));
```
Extract whole `sprintf()` to a variable.

Commit:
[BE] Upload user avatar
https://trello.com/c/pBNMU33u/76-be-upload-user-avatar

1.Extracted sprintf() to a variable
2.Added different catch{} blocks for different exceptions in UserController
